File: src/main/java/dev/psygamer/wireframe/util/collection/FreezableArrayList.java
package dev.psygamer.wireframe.util.collection;

import java.util.ArrayList;
import java.util.Collection;

public class FreezableArrayList <E> extends FreezableList<E> {
	
	public FreezableArrayList() {
		super(new ArrayList<>());
	}
	
	public FreezableArrayList(final int initialCapacity) {
		super(new ArrayList<>(initialCapacity));
	}
	
	public FreezableArrayList(final Collection<? extends E> c) {
		super(new ArrayList<>(c));
	}
	
	@Override
	public FreezableList<E> copy() {
		final FreezableList<E> list = new FreezableArrayList<>(this);
		
		if (isFrozen())
			list.freeze();
		
		return list;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/collection/FreezableArrayList.kt
package dev.psygamer.wireframe.util.collection

class FreezableArrayList<E> : FreezableList<E> {

	constructor() : super(ArrayList<E>())
	constructor(initialCapacity: Int) : super(ArrayList<E>(initialCapacity))
	constructor(elements: Collection<E>) : super(ArrayList<E>(elements))
}